### ISSUE TITLE
[Bug] Firestore reads do not work unless there was a write before

### ISSUE THREAD

**Issue by dex3r** [firebase/quickstart-unity#1154]
### [REQUIRED] Please fill in the following fields:

  * Unity editor version: 2020.3.13f1
  * Firebase Unity SDK version: 8.3.0 (also present on 8.1.0)
  * Source you installed the SDK: .unitypackage
  * Problematic Firebase Component: Firestore
  * Other Firebase Components in use: Firestore
  * Additional SDKs you are using: none
  * Platform you are using the Unity editor on: Windows
  * Platform you are targeting: Android, tested in Editor
  * Scripting Runtime: Editor, scripting backend set to IL2CPP

### [REQUIRED] Please describe the issue here:

On an otherwise empty scene I have one script that reads a document from Firestore. `GetSnapshotAsync` always, 10/10 times throws an exception `FirestoreException: Failed to get document from server.` UNLESS(!) before this read, I execute any write operation on the Firestore (or so it seems).

Once again, the read is ONLY successful if there is a write before it. This is really bizarre.
Also, the initial write always takes a very long time, from my tests it's always 20 seconds.

The strange thing is, earlier today everything was working perfectly. I started working on another project, came back to this one, and my app just stopped working. No changes in git today. I've tried restarting Unity, PC, installed newest SDK (8.3.0). Nothing helped.

#### Steps to reproduce:

1. Run the following code.
2. Exit play mode.
3. Uncomment `TestWrite` call
4. Run the following code again.

This problem is so utterly bizarre that I wonder if this is somehow a problem with Firebase cache on my PC. How can I delete everything that Firebase/Firestore cached on my Windows machine?

I would like to see and attach more logs, however switching logging to Verbose does not seem to do anything in Editor.

#### Relevant Code:

```
public class TestScript : MonoBehaviour
{
    async void Start()
    {
        //await TestWrite(); // The following Read works ONLY if this line is not commented out
        await TestRead();
    }

    private async Task TestRead()
    {
        Debug.Log("Read starting");
        
        DocumentReference doc = FirebaseFirestore.DefaultInstance.Document("test/test");
        await doc.GetSnapshotAsync(Source.Server);

        Debug.Log("Read finished");
    }

    private async Task TestWrite()
    {
        Debug.Log("Write starting");
        
        var loginCounter = FirebaseFirestore.DefaultInstance.Document("test/test");
        var data = new Dictionary<string, object>();

        await loginCounter.SetAsync(data);
        
        Debug.Log("Write finished");
    }
}
```

This gives two possible results, depending if `TestWrite` is commented out:

With write and then read:
![with_write](https://user-images.githubusercontent.com/3155725/136483461-1b5326fd-dc85-4a02-b99e-506057ac5643.png)

Read only:
![no_write](https://user-images.githubusercontent.com/3155725/136483497-745e9694-afd8-444e-a08f-fa8c0578cdba.png)

**Comment by paulinon** [comment]
Hi @dex3r,

Thanks for reporting this issue. In order to be able to replicate this, could you also include the libraries you used in the code you provided? 

Additionally, in the steps you provided, could you add if you've observed any changes in your data from the Firebase console?

**Comment by dconeybe** [comment]
I'll take a look.

**Comment by dconeybe** [comment]
Oh also, could you try to reproduce using the Firestore emulator?

**Comment by dex3r** [comment]
Logs with timestamps: https://pastebin.com/mv2tzdFJ
Firestore emulator works fine.

Also, I realized I'm behind double NAT. This might be the issue here, however, as I said, I have this network configuration for years and everything else works fine.

**Comment by dconeybe** [comment]
Here is the app: https://github.com/dconeybe/UnityIssue1154TestApp
Let me know if you can get it working. Once you confirm, I'll roll a custom build for you with some additional logging.

**Comment by dex3r** [comment]
@dconeybe https://pastebin.com/c2jqX0KL

**Comment by dconeybe** [comment]
Sorry for the long delay. We were heads down on the latest Unity SDK release. Anyways, that is past us now and I have a new build for you to test: https://github.com/firebase/firebase-cpp-sdk/actions/runs/1437861123. Could you run the app with `debug\firebase_unity_issue_1154_test_app.exe write --debug` and post the output at your convenience. Thank you!

**Comment by dex3r** [comment]
https://gist.github.com/dex3r/3e4e5854776669409d338726d40ea277

**Comment by dconeybe** [comment]
Here's another build: https://github.com/firebase/firebase-cpp-sdk/actions/runs/1442420763. You can omit the `--debug` flag this time. At this point it's just adding noise to the logs.

**Comment by dex3r** [comment]
Here you go :)
https://gist.github.com/dex3r/01ec92ed8f79c50389bf93a7cfb823f4

**Comment by dconeybe** [comment]
FYI I'm on vacation for a week so the next build will be delayed. I'm also a bit hung out now adding logging deep in the bowels of grpc, which is proving to be more difficult than I had hoped. Anyways, thanks for your patience.

**Comment by dex3r** [comment]
https://gist.github.com/dex3r/1c86ca2e6e1ccca71d3b5dc0f6936ef0

**Comment by dconeybe** [comment]
Huh, this line from that log file is interesting, especially since it is the first log line after the 30-minute pause:

> Connect failed: {"created":"@1638366677.694000000","description":"socket is null","file":"D:\a\firebase-cpp-sdk\firebase-cpp-sdk\out-sdk\external\src\firestore-build\external\src\grpc\src\core\lib\iomgr\tcp_client_windows.cc","file_line":113}

Here is the line where that error is reported:

https://github.com/grpc/grpc/blob/205d7569eadd03456e2a692a011d630c0ba42a12/src/core/lib/iomgr/tcp_client_windows.cc#L112

I'll add some more logging around that line and get you a new build in the next couple of days.

**Comment by dconeybe** [comment]
Please try out this build that has some logging in `tcp_client_windows.cc`:

https://github.com/firebase/firebase-cpp-sdk/actions/runs/1548128356

**Comment by dex3r** [comment]
https://gist.github.com/dex3r/5187cd936178b94959e094c7318b762f

**Comment by dconeybe** [comment]
And another one: https://github.com/firebase/firebase-cpp-sdk/actions/runs/1550241575

**Comment by dex3r** [comment]
https://gist.github.com/dex3r/385b6a3640dec839b4bae9e8d164b908

**Comment by dconeybe** [comment]
Update: I'm discussing this issue internally with the grpc team. I don't have an ETA for the next build for testing.

**Comment by dconeybe** [comment]
I have another build for testing: https://github.com/firebase/firebase-cpp-sdk/actions/runs/1580732639.

This build has a workaround for a potential race condition that may be causing the 20-second delay that you are experiencing. So it will be very interesting if you still experience the 20-second delay when you test out this build.

**Comment by dex3r** [comment]
https://gist.github.com/dex3r/146d8087ac418dca15785282b0f0cc45

**Comment by dconeybe** [comment]
Dang, that didn't fix the 20-second delay. I'll keep working with the internal teams on this.

**Comment by amoswazana** [comment]
Hi @dconeybe,
I am also experiencing this issue all of a sudden.
Calling db.Document(path).GetSnapshotAsync(Source.Server) will only work if I wait at least 20 (usually around 25-27) seconds from any first call to Firestore.

System information:

- Unity editor version: 2020.3.03f1
- Firebase Unity SDK version: 8.7.0 (also on 8.4.0)
- Source you installed the SDK: .unitypackage
- Problematic Firebase Component: Firestore
- Other Firebase Components in use: None
- Additional SDKs you are using: None
- Platform you are using the Unity editor on: MacBook Pro
- Platform you are targeting: Android and iOS, tested in Editor
- Scripting Runtime: Editor, scripting backend set to either IL2CPP or Mono

If you need me to provide any more information I will be happy to help.

**Comment by dconeybe** [comment]
@amoswazana That's interesting that you're experiencing this on macOS. This issue was reported against Windows and I was working under the assumption that it is a Windows-specific issue. Granted, you could be experiencing a different issue than the OP... but possibly not.

Would you be able to reproduce with the custom testapp that I created and was using with the OP? Just clone https://github.com/dconeybe/UnityIssue1154TestApp, follow the instructions there, and use the C++ SDK build from https://github.com/firebase/firebase-cpp-sdk/actions/runs/1580732639. If you could capture and post the output from running `./firebase_unity_issue_1154_test_app --debug write read` that would be very helpful.

**Comment by amoswazana** [comment]
Tried getting it to work but I seem to have failed.
Here is the output:
https://gist.github.com/amoswazana/574487bbeef11d8a0f8ad967c3a8ab2b

I would also like to add here that yesterday I noticed another issue that happened solely on wifi and not while connected to my mobile hotspot: while building an Android version of our app (with Unity) my gradle process failed reaching https://jcenter.bintray.com with a 502 BAD GATEWAY error code. Building the same app while connected to a hotspot didn't produce any issues.
Today this issue didn't persist but the Firebase one did.
I have sent a query to my ISP to see if something has changed on their end... Very curious indeed.

**Comment by dconeybe** [comment]
@amoswazana Try running `rm -rf ~/Library/Application\ Support/firestore` to delete the cached Firestore data. I'm not sure how but somehow it appears to have become corrupted on your computer. Then, try to run the testapp again.

**Comment by amoswazana** [comment]
OK that solved it. Here is the output to '--debug write read':
https://gist.github.com/amoswazana/d1799a8970b8e66627ece3217c9de1b9

**Comment by dconeybe** [comment]
@amoswazana It looks like you are experiencing the same issue as the OP. I'm going to create a custom C++ SDK build that has some extra logging in it. I had added a bunch of extra logging to Windows-specific grpc files and I'll need to make the same/similar modifications to the macOS-specific grpc files. Once that build is ready I'm hoping you can reproduce with the testapp again.

**Comment by amoswazana** [comment]
Sure thing. Keep me posted and thanks for your quick response 👍

**Comment by dconeybe** [comment]
@amoswazana Could you try reproducing with the testapp again with this build, just like you did before: https://github.com/firebase/firebase-cpp-sdk/actions/runs/1710537244. I've added some macOS-specific logging. It won't give any answers but will guide where I put logging in the next build. Note that you can now omit the `--debug` flag when you run the testapp since it just adds noise to the log output.

**Comment by amoswazana** [comment]
Everything seems to be working now. Both in the cpp and in Unity. The problem solved itself somehow and now we can't reproduce the bug... Just to clarify: nothing changed on my end.
Here is the output of the cpp anyway:
https://gist.github.com/amoswazana/257170296b5fecc14e83d11c1327589e

**Comment by dconeybe** [comment]
@amoswazana I'm glad to hear that this issue has disappeared for you. If it comes back, please run the CPP testapp again and post the logs. I wonder if it's an issue with a specific Firestore endpoint that it's connecting to.

**Comment by dconeybe** [comment]
@dex3r Would you also be able to reproduce with this latest build: https://github.com/firebase/firebase-cpp-sdk/actions/runs/1580732639. The only difference from the last build is that it will log the IP address of the Firestore backend to which it connects.

**Comment by dex3r** [comment]
done, output here: https://gist.github.com/dex3r/9875aaaaf3c1897873604d296bbafef0

**Comment by dconeybe** [comment]
@dex3r My apologies, I gave you the link to the wrong build. That was an old build that is missing the logging that I'm interested in. Could you try again with this build: https://github.com/firebase/firebase-cpp-sdk/actions/runs/1710537244

**Comment by amoswazana** [comment]
Hey guys,
Just to update that I am experiencing what seems to be the same issue once more. Hopefully this will solve itself like it did back in the day.
Is there any other testing you would like me to do while it's still happening?

**Comment by amoswazana** [comment]
Hi @dconeybe,
Unity editor version: 2020.3.28f1
Firebase Unity SDK version: 9.0.0
Platform: Macbook pro.

Let's hope it works itself out...

**Comment by dconeybe** [comment]
Is anyone still experiencing this issue? Apologies that we haven't spent any time on it lately.

**Comment by dex3r** [comment]
@dconeybe Thank you for your engagement in my issue.

Unfortunately, due to a number of bugs and other problems I've encountered during further development I had to move out of Firebase. 

One thing I can recommend to the Firebase team is to move Unity SDK natively to C#. Trying to fix issues with C++ SDK with a cumbersome Swig layer on top is an absolute nightmare.

**Comment by dconeybe** [comment]
Update: This problem could very well have been due to a low-level bug in the grpc library used by Firestore. We have been prevented from upgrading this dependency due to supporting C++11 in https://github.com/firebase/firebase-cpp-sdk (upon which the Unity SDK is built). But soon we will upgrade the minimum-supported version to C++14 and will upgrade the grpc dependency shortly thereafter. Once that happens this problem may go away. One PR to keep eyes on if interested in the grpc upgrade is https://github.com/firebase/firebase-ios-sdk/pull/10650.
